### PR TITLE
Feature/rewrite for client side eval

### PR DESCRIPTION
**Pull request by matthewelwell**
This PR contains a full rewrite of the Python client to allow for client side evaluation. As part of this work, I have also added the possibility for providing a list of default flags when initiating the client. 

Some general design decisions: 

 1. The client only exposes 2 methods, one to get an environments flags and another to get the flags for an identity (which also can optionally persist a list of traits). 
 2. The client returns a Flags class, from which, users can access specific flags and their state / value. This means that users will have to write a little more code to get the value of a given flag, however, it will only make a single API request to do so (for users not using client side evaluation). 

Options for extension: 

1. Caching
2. At the moment, defaults are only used when the API request is successful but the feature doesn't exist. We should extend this to allow the defaults to be used when the API requests are failing.

### CODE REVIEW

File: flagsmith/flagsmith.py [hunk 1]
```diff
@@ -1,219 +1,212 @@
...
+
+    def all_flags(self) -> typing.List[Flag]:
+        return list(self.flags.values())
+
+    def get_flag(self, feature_name: str) -> typing.Optional[Flag]:
+        return self.flags.get(feature_name)
 
 
-class Flagsmith:
+class FlagsmithV3:
```
Comment (dabeeeenster) [inline flagsmith/flagsmith.py:83]: Why v3?

File: flagsmith/flagsmith.py [hunk 2]
```diff
@@ -1,219 +1,212 @@
...
     def __init__(
-        self, environment_id, api=SERVER_URL, custom_headers=None, request_timeout=None
+        self,
+        environment_key: str,
+        api_url: str = API_URL,
+        custom_headers: typing.Dict[str, typing.Any] = None,
+        request_timeout: int = None,
+        enable_client_side_evaluation: bool = False,
+        environment_refresh_interval_seconds: int = 10,
```
Comment (dabeeeenster) [inline flagsmith/flagsmith.py:91]: Think this should default to 60?

File: flagsmith/flagsmith.py [hunk 3]
```diff
@@ -1,219 +1,212 @@
...
+    def all_flags(self) -> typing.List[Flag]:
+        return list(self.flags.values())
+
+    def get_flag(self, feature_name: str) -> typing.Optional[Flag]:
+        return self.flags.get(feature_name)
 
 
-class Flagsmith:
+class FlagsmithV3:
     def __init__(
```
Comment (dabeeeenster) [inline flagsmith/flagsmith.py:84]: Think we would want a flag to be able to disable the flag analytics as well

**Comment by matthewelwell** [comment]
Design decisions to discuss: 

 - The methods on the flagsmith client return a new class called `Flags`, it is here that the user can interact with the flags and determine their enabled status / value. This logically ties in with the API requests - user makes one API request and subsequently can get the information for multiple flags.

File: flagsmith/models.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        """
+        Get the value of a particular feature.
+
+        :param feature_name: the name of the feature to retrieve the value of.
+        :return: the value of the given feature.
+        :raises FlagsmithClientError: if feature doesn't exist
+        """
+        return self.get_flag(feature_name).value
+
+    def get_flag(self, feature_name: str) -> typing.Optional[BaseFlag]:
```
[the remark this answers is not in the document]
Comment (matthewelwell) [reply flagsmith/models.py:120]: Yep, good point. I guess one question is whether we should raise the exception or return None if it's not found.

[the remark this answers is not in the document]
Comment (gagantrivedi) [reply flagsmith/models.py:120]: I think raising the error is more pythonic?

[the remark this answers is not in the document]
Comment (matthewelwell) [reply flagsmith/models.py:120]: Yep, you're right. Let's see if anyone complains. I'll update the typehint now.

**Review by gagantrivedi — APPROVED** [review]

**Pull request merged**